### PR TITLE
Add upickle module

### PR DESCRIPTION
**Pull request by lolgab**
- [x] ~Blocked by the release of https://github.com/lihaoyi/upickle/pull/326~ Changed the imports to not face the bug

Before submitting pull request:
- [x] Check if the project compiles by running `sbt compile`
- [x] Verify docs compilation by running `sbt compileDocs`
- [x] Check if tests pass by running `sbt test`
- [x] Format code by running `sbt scalafmt`

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -215,9 +215,9 @@ val compileAndTest = "compile->compile;test->test"
 def dependenciesFor(version: String)(deps: (Option[(Long, Long)] => ModuleID)*): Seq[ModuleID] =
   deps.map(_.apply(CrossVersion.partialVersion(version)))
 
-lazy val coreProjectAggregates: Seq[ProjectReference] = if (sys.env.isDefinedAt("STTP_NATIVE")) {
+lazy val projectAggregatesWithNative: Seq[ProjectReference] = if (sys.env.isDefinedAt("STTP_NATIVE")) {
```
Comment (adamw) [inline build.sbt:218]: this contains the native projects optionally, so maybe `projectWithOptionalNative` or sth like that?

File: build.sbt [hunk 2]
```diff
@@ -345,13 +351,7 @@ lazy val core = (projectMatrix in file("core"))
     scalaVersions = List(scala2_11),
     settings = {
       commonNativeSettings ++ intellijSkipImport ++ List(
-        libraryDependencies ++= Seq(
-          "org.scala-native" %%% "test-interface" % scalaNativeTestInterfaceVersion % Test,
```
Comment (adamw) [inline build.sbt:349]: so we don't need the test-interface anymore?

**Comment by adamw** [comment]
Looks good - we're waiting for the upickle update then :)

File: build.sbt (hunk 2 above)
Comment (lolgab) [reply build.sbt:349]: It is automatically added to the `Test` scope by the Scala Native sbt plugin.
Verified by running `show upickleNative2_11/Test/libraryDependencies` which lists it.

File: build.sbt (hunk 1 above)
Comment (lolgab) [reply build.sbt]: I renamed to `projectsWithOptionalNative` since they are more than one :)

**Comment by lolgab** [comment]
At the end I didn't want to wait for the upickle release and worked around the issue by not using the wildcard import in the tests.
If you want I can do a follow up PR once a new version of upickle is released.

**Comment by adamw** [comment]
So travis worked in the end but now I'm getting

```
[error] java.nio.file.ClosedFileSystemException
[error] 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.ensureOpen(ZipFileSystem.java:1112)
[error] 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.newByteChannel(ZipFileSystem.java:694)
[error] 	at jdk.zipfs/jdk.nio.zipfs.ZipPath.newByteChannel(ZipPath.java:815)
[error] 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.newByteChannel(ZipFileSystemProvider.java:252)
[error] 	at java.base/java.nio.file.Files.newByteChannel(Files.java:370)
[error] 	at java.base/java.nio.file.Files.newByteChannel(Files.java:421)
[error] 	at java.base/java.nio.file.Files.readAllBytes(Files.java:3205)
[error] 	at scala.scalanative.io.VirtualDirectory$NioDirectory.read(VirtualDirectory.scala:72)
[error] 	at scala.scalanative.io.VirtualDirectory$NioDirectory.read$(VirtualDirectory.scala:71)
[error] 	at scala.scalanative.io.VirtualDirectory$JarDirectory.read(VirtualDirectory.scala:97)
[error] 	at scala.scalanative.linker.ClassPath$Impl.$anonfun$load$2(ClassPath.scala:50)
[error] 	at scala.Option.map(Option.scala:230)
[error] 	at scala.scalanative.linker.ClassPath$Impl.$anonfun$load$1(ClassPath.scala:49)
[error] 	at scala.collection.mutable.HashMap.getOrElseUpdate(HashMap.scala:86)
[error] 	at scala.scalanative.linker.ClassPath$Impl.load(ClassPath.scala:49)
[error] 	at scala.scalanative.linker.ClassLoader$FromDisk$$anonfun$load$1.applyOrElse(ClassLoader.scala:29)
[error] 	at scala.scalanative.linker.ClassLoader$FromDisk$$anonfun$load$1.applyOrElse(ClassLoader.scala:27)
[error] 	at scala.collection.TraversableOnce.collectFirst(TraversableOnce.scala:172)
[error] 	at scala.collection.TraversableOnce.collectFirst$(TraversableOnce.scala:159)
[error] 	at scala.collection.AbstractTraversable.collectFirst(Traversable.scala:108)
[error] 	at scala.scalanative.linker.ClassLoader$FromDisk.load(ClassLoader.scala:27)
[error] 	at scala.scalanative.linker.Reach.lookup(Reach.scala:64)
[error] 	at scala.scalanative.linker.Reach.reachDefn(Reach.scala:88)
[error] 	at scala.scalanative.linker.Reach.reachGlobalNow(Reach.scala:158)
[error] 	at scala.scalanative.linker.Reach.classInfo(Reach.scala:304)
[error] 	at scala.scalanative.linker.Reach.reachOp(Reach.scala:532)
[error] 	at scala.scalanative.linker.Reach.reachInst(Reach.scala:474)
[error] 	at scala.scalanative.linker.Reach.$anonfun$reachInsts$1(Reach.scala:468)
[error] 	at scala.scalanative.linker.Reach.$anonfun$reachInsts$1$adapted(Reach.scala:468)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:943)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:943)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[error] 	at scala.scalanative.linker.Reach.reachInsts(Reach.scala:468)
[error] 	at scala.scalanative.linker.Reach.reachDefine(Reach.scala:397)
[error] 	at scala.scalanative.linker.Reach.reachDefn(Reach.scala:113)
[error] 	at scala.scalanative.linker.Reach.$anonfun$reachDefn$2(Reach.scala:94)
[error] 	at scala.scalanative.linker.Reach.$anonfun$reachDefn$2$adapted(Reach.scala:90)
[error] 	at scala.Option.fold(Option.scala:251)
[error] 	at scala.scalanative.linker.Reach.reachDefn(Reach.scala:90)
[error] 	at scala.scalanative.linker.Reach.process(Reach.scala:82)
[error] 	at scala.scalanative.linker.Reach$.apply(Reach.scala:673)
[error] 	at scala.scalanative.linker.Link$.$anonfun$apply$1(Link.scala:12)
[error] 	at scala.scalanative.util.Scope$.apply(Scope.scala:30)
[error] 	at scala.scalanative.linker.Link$.apply(Link.scala:11)
[error] 	at scala.scalanative.build.ScalaNative$.$anonfun$link$1(ScalaNative.scala:34)
[error] 	at scala.scalanative.build.Logger.time(Logger.scala:32)
[error] 	at scala.scalanative.build.Logger.time$(Logger.scala:29)
[error] 	at scala.scalanative.build.Logger$$anon$2.time(Logger.scala:66)
[error] 	at scala.scalanative.build.ScalaNative$.link(ScalaNative.scala:34)
[error] 	at scala.scalanative.build.Build$.$anonfun$build$1(Build.scala:55)
[error] 	at scala.scalanative.build.Logger.time(Logger.scala:32)
[error] 	at scala.scalanative.build.Logger.time$(Logger.scala:29)
[error] 	at scala.scalanative.build.Logger$$anon$2.time(Logger.scala:66)
[error] 	at scala.scalanative.build.Build$.build(Build.scala:53)
[error] 	at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.$anonfun$scalaNativeConfigSettings$11(ScalaNativePluginInternal.scala:145)
[error] 	at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.interceptBuildException(ScalaNativePluginInternal.scala:220)
[error] 	at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.$anonfun$scalaNativeConfigSettings$10(ScalaNativePluginInternal.scala:145)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] 	at java.base/java.lang.Thread.run(Thread.java:834)
[error] (upickleNative2_11 / Nativetest / nativeLink) java.nio.file.ClosedFileSystemException
```

during the release build. Looks like it needs investigating what's happening there :)

**Comment by lolgab** [comment]
It can be related to a bug with aggregate projects on Scala Native. I couldn't reproduce by the way..
I think that having an alias that tests / publishes only the native sub-projects could work around this problem \(which [should be solved in Scala Native master](https://gitter.im/scala-native/scala-native?at=5f9c5903c10273610ad6988f)\).
But it is ugly :(

**Comment by adamw** [comment]
[This](https://github.com/softwaremill/sttp/commit/6ec13ab8c57714830ea8220f54f682784cb561f8) hack worked ;)

Released in https://github.com/softwaremill/sttp/releases/tag/v3.0.0-RC8

**Comment by lolgab** [comment]
Nice! :)
Thank you and sorry for wasting your time having to find hacks!

**Comment by adamw** [comment]
Oh, no worries, it's expected when you are using libraries in version not only 0.x but 0.4.0-M2 ;-)

**Pull request merged**